gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/reactive/handlers/api/security/AbstractSecurityChain.java

```diff
@@ -0,0 +1,135 @@
+ */
+@Slf4j
+public abstract class AbstractSecurityChain<
+    P extends AbstractSecurityPlan<? extends BaseSecurityPolicy, C>, C extends BaseExecutionContext
+> {
+
+    protected static final String PLAN_UNRESOLVABLE = "GATEWAY_PLAN_UNRESOLVABLE";
+    protected static final String PLAN_RESOLUTION_FAILURE = "GATEWAY_PLAN_RESOLUTION_FAILURE";
+    protected static final String UNAUTHORIZED_MESSAGE = "Unauthorized";
+    protected static final String TEMPORARILY_UNAVAILABLE_MESSAGE = "Temporarily Unavailable";
+    protected static final String ATTR_INTERNAL_PLAN_RESOLUTION_FAILURE = "securityChain.planResolutionFailure";
+
+    protected static final Single<Boolean> TRUE = Single.just(true);
+    protected static final Single<Boolean> FALSE = Single.just(false);
+    private final Flowable<P> chain;
+
+    public AbstractSecurityChain(Flowable<P> securityPlans) {
+        this.chain = securityPlans;
+    }
+
+    protected abstract Completable sendError(C ctx, ExecutionFailure failure);
+
+    protected abstract SingleSource<Boolean> executePlan(P securityPlan, C ctx);
+
+    /**
+     * Executes the security chain by executing all the {@link SecurityPlan}s in an ordered sequence.
+     * It's up to each {@link SecurityPlan} to provide its order. The lower is the order, the highest priority is.
+     * The result of the security chain execution depends on the first {@link SecurityPlan} able to execute the request.
+     * If no {@link SecurityPlan} has been executed because there is no {@link SecurityPlan} in the chain or none of them can execute the request,
+     * then the chain is interrupted with a 401 response status and the {@link Completable} returns an error.
+     *
+     * @param ctx the current execution context.
+     * @return a {@link Completable} that completes if the request has been successfully handled by a {@link SecurityPlan} or returns
+     * an error if no {@link SecurityPlan} can execute the request or the {@link SecurityPlan} failed.
+     */
+    public Completable execute(C ctx) {
+        return defer(() -> {
+            if (!Objects.equals(true, ctx.getInternalAttribute(ATTR_INTERNAL_SECURITY_SKIP))) {
+                return chain
+                    .concatMapSingle(securityPlan -> continueChain(ctx, securityPlan))
+                    .any(Boolean::booleanValue)
+                    .flatMapCompletable(securityHandled -> {
+                        if (Boolean.FALSE.equals(securityHandled)) {
+                            Throwable throwable = ctx.getInternalAttribute(ATTR_INTERNAL_PLAN_RESOLUTION_FAILURE);
+                            if (throwable != null) {
+                                return sendError(
+                                    ctx,
+                                    new ExecutionFailure(SERVICE_UNAVAILABLE_503)
+                                        .key(PLAN_RESOLUTION_FAILURE)
+                                        .message(TEMPORARILY_UNAVAILABLE_MESSAGE)
```

Comment:
as discussed, in the future we might want to specialize the ExecutionFailure but it it is ok to keep it simple here